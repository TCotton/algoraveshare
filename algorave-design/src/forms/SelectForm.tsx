import * as Ariakit from "@ariakit/react";
import {SelectProvider} from "@ariakit/react";

type SelectFormProps = {
    label: string;
    items: { label: string; value: string }[];
};

export default function SelectForm(props: SelectFormProps) {
    // Ariakit expects items to have id, label, value
    const items = props.items.map((item, idx) => ({
        id: item.value,
        label: item.label,
        value: item.value
    }));
    const select = Ariakit.useSelectStore({items});
    return (
        <div className="select-container">
            <SelectProvider>
                <Ariakit.SelectLabel store={select} className="select-label">
                    {props.label}
                </Ariakit.SelectLabel>
                <Ariakit.Select store={select} className="select-trigger">
                    <Ariakit.SelectValue/>
                    <Ariakit.SelectArrow className="select-arrow"/>
                </Ariakit.Select>
                <Ariakit.SelectPopover store={select} className="select-popover">
                    {items.map(item => (
                        <Ariakit.SelectItem className="select-item" value={item.value} key={item.id}>
                            {item.value}
                        </Ariakit.SelectItem>
                    ))}
                </Ariakit.SelectPopover>
            </SelectProvider>
        </div>
    );
};