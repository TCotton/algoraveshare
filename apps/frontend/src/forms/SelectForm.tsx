/* eslint-disable react/no-unknown-property */
import React from 'react'
import * as Ariakit from '@ariakit/react'
import { SelectProvider } from '@ariakit/react'
import type { SelectFormProps } from './formtypes.ts'

interface FormObject {
  useValue: (name: string) => string
  setValue: (name: string, value: string) => void
}

/**
 * In the future use the data-test-id for PlayWright tests:
 *
 * SelectForm
 *   label="Choose the project software"
 *   name="projectSoftware"
 *   form={form}
 *   items={[...]}
 *   selectClass="project-software"
 *   data-test-id="project-software-select"
 * />
 *
 *
 * // Instead of using class selectors:
 * await page.locator('.project-software button.select-trigger').click()
 *
 * // You can now use data-test-id:
 * await page.locator('[data-test-id="project-software-select"] button.select-trigger').click()
 *
 * // Or more simply:
 * await page.getByTestId('project-software-select').click()
 */

interface ExtendedSelectFormProps extends SelectFormProps {
  'name': string
  'form'?: FormObject
  'selectClass'?: string
  'onChange'?: (name: string, value: string) => void
  'onMouseEnter'?: (event: React.MouseEvent<HTMLDivElement>) => void
  'onMouseLeave'?: (event: React.MouseEvent<HTMLDivElement>) => void
  'data-testid'?: string
}

export default function SelectForm(props: ExtendedSelectFormProps) {
  const { items, label, name, form, onChange, selectClass, onMouseEnter, onMouseLeave, 'data-testid': dataTestId } = props
  const formattedItems = items?.map((item: { value: string, label: string }) => ({
    id: item.value,
    label: item.label,
    value: item.value,
  }))
  const select = Ariakit.useSelectStore({
    defaultValue: items.length > 0 ? items[0].value : undefined,
    items: formattedItems,
    value: form ? form.useValue(name) : undefined,
    setValue: (val: string) => {
      if (form) {
        form.setValue(name, val)
      }
      if (onChange) {
        onChange(name, val)
      }
    },
  })
  return (
    <div
      className={`select-container ${selectClass || ''}`}
      is-="typography-block"
      box-="round"
      shear-="top"
      onMouseEnter={onMouseEnter}
      onMouseLeave={onMouseLeave}
      data-testid={dataTestId}
    >
      <SelectProvider store={select}>
        <Ariakit.SelectLabel store={select} className="select-label">
          <div is-="badge" variant-="background0">{label}</div>
        </Ariakit.SelectLabel>
        <Ariakit.Select store={select} className="select-trigger">
          <Ariakit.SelectValue fallback={items.length > 0 ? items[0].value : undefined} />
          <Ariakit.SelectArrow className="select-arrow" />
        </Ariakit.Select>
        <Ariakit.SelectPopover store={select} className="menu-wrapper" id="menu-form" gutter={4}>
          {formattedItems.map((item: { value: string, label: string, id: string }) => (
            <Ariakit.SelectItem className={`menu-item ${item.label}`} value={item.value} key={item.id}>
              <a href="#menu-form">{item.value}</a>
            </Ariakit.SelectItem>
          ))}
        </Ariakit.SelectPopover>
      </SelectProvider>
    </div>
  )
}
