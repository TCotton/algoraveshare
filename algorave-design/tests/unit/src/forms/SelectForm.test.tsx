import React from "react";
import {describe, it, expect, beforeEach, vi} from "vitest";
import {render, screen} from "@testing-library/react";
import renderer from 'react-test-renderer';
import SelectForm from "../../../../src/forms/SelectForm";

vi.mock("@ariakit/react", () => ({
    __esModule: true,
    useSelectStore: vi.fn(() => ({})),
    Select: (props: any) => <select {...props}>{props.children}</select>,
    SelectItem: (props: any) => <option {...props}>{props.children}</option>,
    SelectPopover: (props: any) => <div {...props}>{props.children}</div>,
    SelectProvider: (props: any) => <div {...props}>{props.children}</div>,
    SelectLabel: (props: any) => <label {...props}>{props.children}</label>,
    SelectValue: () => <span>Selected Value</span>,
    SelectArrow: () => <span>â–¼</span>
}));

// Sample data
const items = {
    "label": "Choose a project type",
    "items": [
        { "value": "Finished Project", "label": "finished" },
        { "value": "Before and After Live Coding Project", "label": "before-after" }
    ]
};

beforeEach(() => {
    render(<SelectForm label={items.label} items={items.items} />);
})

describe("SelectForm", () => {
    it("renders label and options", () => {
        expect(screen.getByText(/choose a project type/i)).toBeInTheDocument();
        expect(screen.getByText("Finished Project")).toBeInTheDocument();
        expect(
            screen.getByText("Before and After Live Coding Project")
        ).toBeInTheDocument();
    });
    it('matches snapshot', () => {
        const tree = renderer.create(<SelectForm label={items.label} items={items.items} />).toJSON();
        expect(tree).toMatchSnapshot();
    })
});
