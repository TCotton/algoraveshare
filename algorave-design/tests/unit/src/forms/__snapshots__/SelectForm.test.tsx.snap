// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SelectForm > matches snapshot 1`] = `null`;
